azerothcore/azerothcore-wotlk
# fix(DB/creature): added sql procedure to remove npc duplicates

## Changes Proposed:
-  added sql procedure to remove npc duplicates based on position, map and events and have a precision loss on position.z of 0.1
-  deletes around ~486 entries from `creature` and `game_event_creature` tables 

## Issues Addressed:

This should fix all the reports about NPC duplication, but **NOT** the missing gameobjects (maybe we should split them in separate tasks).
- https://github.com/azerothcore/azerothcore-wotlk/issues/6469
- https://github.com/chromiecraft/chromiecraft/issues/946
- https://github.com/chromiecraft/chromiecraft/issues/957
- https://github.com/chromiecraft/chromiecraft/issues/950
- https://github.com/chromiecraft/chromiecraft/issues/948
- https://github.com/chromiecraft/chromiecraft/issues/945
- https://github.com/chromiecraft/chromiecraft/issues/944
- https://github.com/chromiecraft/chromiecraft/issues/943
- https://github.com/chromiecraft/chromiecraft/issues/942
- https://github.com/chromiecraft/chromiecraft/issues/940
- https://github.com/chromiecraft/chromiecraft/issues/936
- https://github.com/chromiecraft/chromiecraft/issues/937

## SOURCE:

Because there are so many NPCs we need to automate the process of deletion if we want to use only 1 NPC for all the events but with different model-ids.

## Tests Performed:

- Win 10 Docker Setup from https://github.com/azerothcore/azerothcore-wotlk/commit/0f57da93008d20f61d212aac493b48ace35de0e4
- Run procedure
- Check npcs in world
- I used all the test data provided by the contributors from https://github.com/chromiecraft/chromiecraft/issues/935 as an `where guid in ( contributor-test-data )`  80% of the reported GUIDs are identified (see issues for the remaining 20%)



## How to Test the Changes:


1. Make a backup of the `creature` and `game_event_creature` tables before running the procedure to avoid installing the whole DB docker volume with `create table creature_bk as select * from from creature` and `create table game_event_creature_bk as select * from from game_event_creature`
2. Copy paste the procedure in a SQL client
3. Run it and start verifying the GUIDs reported

## Known Issues and TODO List:

- The problem is there are some small precisions for `position.x` and `position.y` and some of the GUIDs reported by contributors are not found by this procedure and adding similar conditions to the group-by will actually return 0 results
- Some NPCs are not found as reported on my DB and I assume because there are different GUIDs between ChromieCraft reports and AzerothCore DB generated from sql-revisions (it's enough for a GM to just do a creature add ingame and the whole guids list will not match anymore)
- I think not all the features used in this procedure are supported by all the mysql versions
- I am not sure if there is any SQL code-convention provided, feel free to add any suggestion about formatting, I used CLION and the database client plugin for formatting
- The procedure takes around ~1.5 minute, added a fixme, maybe I should add some indexes ?
- Some of the builds are failing, should I delete from `creature_addon` table too ?
- Other builds are failing with exit code 1 and a semicolon check, not sure what it means
- We could add a LIMIT to the delete query to avoid magically deleting the entire tables
- Should I add the `smart_scripts` table into the inner joins for the creatures ?


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
